app/scripts/controllers/permissions/specifications.js

```diff
@@ -0,0 +1,261 @@
+import {
+  CaveatTypes,
+  RestrictedMethods,
+} from '../../../../shared/constants/permissions';
+
+/**
+ * This file contains the specifications of the permissions and caveats
+ * that are recognized by our permission system. See the PermissionController
+ * README in @metamask/snap-controllers for details.
+ */
+
+/**
+ * The "keys" of all of permissions recognized by the PermissionController.
+ * Permission keys and names have distinct meanings in the permission system.
+ */
+const PermissionKeys = Object.freeze({
+  ...RestrictedMethods,
+});
+
+/**
+ * Factory functions for all caveat types recognized by the
+ * PermissionController.
+ */
+const CaveatFactories = Object.freeze({
+  [CaveatTypes.restrictReturnedAccounts]: (accounts) => {
+    return { type: CaveatTypes.restrictReturnedAccounts, value: accounts };
+  },
+});
+
+/**
+ * A PreferencesController identity object.
+ *
+ * @typedef {Object} Identity
+ * @property {string} address - The address of the identity.
+ * @property {string} name - The name of the identity.
+ * @property {number} [lastSelected] - Unix timestamp of when the identity was
+ * last selected in the UI.
+ */
+
+/**
+ * Gets the specifications for all caveats that will be recognized by the
+ * PermissionController.
+ *
+ * @param {{
+ *   getIdentities: () => Record<string, Identity>,
+ * }} options - Options bag.
+ */
+export const getCaveatSpecifications = ({ getIdentities }) => {
+  return {
+    [CaveatTypes.restrictReturnedAccounts]: {
+      type: CaveatTypes.restrictReturnedAccounts,
+
+      decorator: (method, caveat) => {
+        return async (args) => {
+          const result = await method(args);
+          return result
+            .filter((account) => caveat.value.includes(account))
+            .slice(0, 1);
+        };
+      },
+
+      validator: (caveat, _origin, _target) =>
+        validateCaveatAccounts(caveat.value, getIdentities),
+    },
+  };
+};
+
+/**
+ * Gets the specifications for all permissions that will be recognized by the
+ * PermissionController.
+ *
+ * @param {{
+ *   getKeyringAccounts: () => Promise<string[]>,
+ *   getIdentities: () => Record<string, Identity>,
+ * }} options - Options bag.
+ */
+export const getPermissionSpecifications = ({
+  getKeyringAccounts,
+  getIdentities,
+}) => {
+  return {
+    [PermissionKeys.eth_accounts]: {
+      targetKey: PermissionKeys.eth_accounts,
+      allowedCaveats: [CaveatTypes.restrictReturnedAccounts],
+
+      factory: (permissionOptions, requestData) => {
+        if (Array.isArray(permissionOptions.caveats)) {
+          throw new Error(
+            `${
+              PermissionKeys.eth_accounts
+            } error: Received unexpected caveats:\n${JSON.stringify(
+              permissionOptions.caveats,
+              null,
+              2,
+            )}`,
+          );
+        }
+
+        // This value will be further validated as part of the caveat.
+        if (!requestData.approvedAccounts) {
```

Comment:
Should we also validate that this is an array of strings? Or is that validation performed elsewhere already?

Reply:
Oh, I guess this is taken care of by the caveat validator. Nevermind then.